### PR TITLE
Prohibit creating objects without properties

### LINKED ISSUE #1943: Validate all ObjectSchemas have >= 1 column
Ensure that every table has at least one property or back link column, otherwise Core won't be able to track inserted "rows" after re-opening the Realm.

### PR DESCRIPTION
**Pull request by mrackwitz**
Fixes #1943.

/c @jpsim @tgoyne

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,7 +7,7 @@ x.x.x Release notes (yyyy-MM-dd)
 
 ### Enhancements
 
-* None.
+* Prohibit to create or add objects of a class without properties to a Realm.
```
Comment (jpsim) [inline CHANGELOG.md]: Maybe this should go in the "Bugfixes" section? And it'd be good to elaborate here:

> Prohibit creating or adding objects of a class with no properties to a Realm. This previously would have silently failed to persist the object beyond the current transaction.

File: Realm/RLMObjectStore.mm [hunk 2]
```diff
@@ -91,6 +91,23 @@ static inline void RLMVerifyInWriteTransaction(__unsafe_unretained RLMRealm *con
...
+    RLMObjectSchema *schema = [realm.schema schemaForClassName:objectClassName];
+    if (!schema) {
+        @throw RLMException(@"Object type '%@' is not persisted in the Realm. "
+                            @"If using a custom `objectClasses` / `objectTypes` array in your configuration, "
+                            @"add `%@` to the list of `objectClasses` / `objectTypes`.",
+                            objectClassName, objectClassName);
+    }
+    if (schema.properties.count == 0) {
+        @throw RLMException(@"Can't add objects of type '%@' to the Realm as the class doesn't "
+                            @"define any persisted properites.",
```
Comment (jpsim) [inline Realm/RLMObjectStore.mm]: managed properties

File: Realm/RLMObjectStore.mm [hunk 3]
```diff
@@ -91,6 +91,23 @@ static inline void RLMVerifyInWriteTransaction(__unsafe_unretained RLMRealm *con
     }
 }
 
+static inline RLMObjectSchema *RLMGetVerifiedObjectSchema(__unsafe_unretained RLMRealm *const realm,
+                                                          NSString *const objectClassName) {
+    RLMObjectSchema *schema = [realm.schema schemaForClassName:objectClassName];
+    if (!schema) {
+        @throw RLMException(@"Object type '%@' is not persisted in the Realm. "
```
Comment (jpsim) [inline Realm/RLMObjectStore.mm]: not managed by the Realm

Comment (mrackwitz) [inline Realm/RLMObjectStore.mm]: #3657 is still not merged yet.

Comment (jpsim) [inline Realm/RLMObjectStore.mm]: we've still settled on that naming and use it in a number of places already.

File: CHANGELOG.md (hunk 1 above)
Comment (mrackwitz) [inline CHANGELOG.md]: :+1:

File: Realm/RLMObjectStore.mm (hunk 3 above)
Comment (mrackwitz) [inline Realm/RLMObjectStore.mm]: The same as in https://github.com/realm/realm-cocoa/pull/3651#discussion_r64659667 should apply here. This PR should just do functional changes in isolation.

**Comment by mrackwitz** [comment]
This needs to be reviewed. /c @austinzheng

File: Realm/RLMObjectStore.mm [hunk 4]
```diff
@@ -91,6 +91,23 @@ static inline void RLMVerifyInWriteTransaction(__unsafe_unretained RLMRealm *con
...
+    RLMObjectSchema *schema = [realm.schema schemaForClassName:objectClassName];
+    if (!schema) {
+        @throw RLMException(@"Object type '%@' is not managed in the Realm. "
+                            @"If using a custom `objectClasses` / `objectTypes` array in your configuration, "
+                            @"add `%@` to the list of `objectClasses` / `objectTypes`.",
+                            objectClassName, objectClassName);
+    }
+    if (schema.properties.count == 0) {
+        @throw RLMException(@"Can't add objects of type '%@' to the Realm as the class doesn't "
+                            @"define any managed properites.",
```
Comment (austinzheng) [inline Realm/RLMObjectStore.mm]: "properties"

File: Realm/Tests/ObjectTests.m [hunk 5]
```diff
@@ -856,6 +856,15 @@ - (void)testAddingObjectNotInSchemaThrows {
     [realm cancelWriteTransaction];
 }
 
+- (void)testAddingObjectWithoutAnyPropertiesThrows {
+    RLMRealm *realm = [RLMRealm defaultRealm];
+
+    [realm beginWriteTransaction];
+    RLMAssertThrowsWithReasonMatching([realm addObject:[[AbstractObject alloc] initWithValue:@[]]], @"Can't add objects of type 'AbstractObject' to the Realm as the class doesn't define any persisted properites.");
```
Comment (austinzheng) [inline Realm/Tests/ObjectTests.m]: Spelling

**Comment by austinzheng** [comment]
👍 , but please rebase first

**Comment by jpsim** [comment]
👍 pending rebase

**Comment by jpsim** [comment]
This is now also being added to core in realm/realm-core#1924, so I think we should close this.

**Comment by mrackwitz** [comment]
Agreed, this seems to be a preferable solution.

**Pull request closed without merging**